### PR TITLE
WIP: Undo-redo for intervals

### LINKED ISSUE #11269: Intervals don't seem to go back to their previous positions after an undo operation
If an interval's start and end offset is @15 in the document and if user deletes the text from @[10 - 20] and does an undo immediately. And now if the interval's start and end offset is queried, it will be @20 instead of @15. 

The ask from this bug is to make interval positions undo friendly. The intervals should be back at their previous position instead of being at the end of inserted text after an undo operation. In the above example, after the undo operation the interval's start and end offset should be @15 instead of @20.

The following are the API being used to created intervals, delete text and query the interval position

APIs to create intervals

> const sharedIntervalCollection = sharedString.getIntervalCollection(intervalType);
> const sharedInterval = sharedIntervalCollection.add(startOffset, endOffset, IntervalType.SlideOnRemove, properties);

APIs to delete text

> sharedString.removeText(startOffset, LengthOfDeletedText);

APIs to get interval position

> const sharedIntervalCollection = sharedString.getIntervalCollection(intervalType);
> const sharedInterval = sharedIntervalCollection.getIntervalById(intervalId);
> To get start position: sharedInterval.start.toPosition()
> To get end position: sharedInterval.end.toPosition()

### PR DESCRIPTION
**Pull request by Abe27342**
## Description

This adds an API to `@fluidframework/sequence` which enables recovering an interval from one of its endpoints. The specific API retrieves the label for the interval collection that endpoint is a part of, as well as the `id` of the interval. This required storing an additional prop on each endpoint reference. I also fixed a bug in treating the label field differently on reloaded intervals.

The motivation for this change is to enable reasonable undo-redo behaviors for SlideOnRemove intervals to be implementable. There is some discussion of the issue on #11269. To that end, I also augmented the example undo-redo handler we have implemented for sequence to demonstrate how one might go about reverting refs. A real implementation would likely filter which refs to restore based on their properties.

One alternative approach I could think of would be to push elements onto the undo-redo stack when the `changeInterval` event is emitted from an interval collection. This is only remotely plausible once the changes from #10966 are released (currently in next). However, that approach suffers from a few timing-related issues: the interval change due to slide won't be emitted until the remote is acked, which is awkward for app code to handle (e.g. one might want to undo an operation before it's acked), and also means we'd need to expose additional APIs on sequence to resolve the correct "previous position": the `previousInterval` argument on the change event only provides a ReferencePosition that lives on a removed segment, but the interval collection APIs operate in terms of positions. So one would need to find the "equivalent position on the newly inserted segments," which would need something like a rebase op. A sample of how that might look (without getting the rebase logic correct) can be found [here](https://github.com/Abe27342/FluidFramework/commit/016064d88cbca54c4d0990bd907ea04e8e5ce03d).

Going with this approach closes #11269.

### CODE REVIEW

File: packages/framework/undo-redo/src/sequenceHandler.ts [hunk 1]
```diff
@@ -85,10 +89,22 @@ export class SharedSegmentSequenceRevertible implements IRevertible {
...
+                }
             }
         }
     }
 
+    private addRemovedRefs(current: ITrackedSharedSegmentSequenceRevertible, segment: ISegment): void {
+        const segRefs = Array.from(segment.localRefs ?? [], (ref) => ({ ref, offset: ref.getOffset() }));
+        if (segRefs.length > 0) {
+            current.refsToRestore ??= new Map();
+            current.refsToRestore.set(segment, segRefs);
```
Comment (anthony-murphy) [inline packages/framework/undo-redo/src/sequenceHandler.ts:104]: i don't think this works, as the segment could still split after being removed, as it is still above the min seq, that could then mess with the mapping.

Comment (Abe27342) [reply packages/framework/undo-redo/src/sequenceHandler.ts:104]: ah, yeah--that makes sense. It should be possible to do something similar but using tracking groups, right? (haven't looked too much into the details, but I see that they split how one would want on segment split)

Then--is that logic something worth exporting from MergeTree? At first glance it looks like it should be possible to write in a tree-shakeable way, so hopefully no impact for non-users.

Comment (Abe27342) [reply packages/framework/undo-redo/src/sequenceHandler.ts:104]: alternatively, seems like `mergeTreeMaintenanceCallback` has enough information to remedy this. It's a lot of burden to put on the implementer though.

Comment (Abe27342) [reply packages/framework/undo-redo/src/sequenceHandler.ts:103]: I pushed a change with a test for this kind of thing (and haven't tweaked implementation, so it fails). Going to table the undo/redo part of this PR for now and split out the interval collection changes into a separate thing.

**Comment by Abe27342** [comment]
Going to close this without merging for now, may re-open later with improved undo-redo or just start fresh. I split out the necessary changes to interval collections elsewhere.

**Pull request closed without merging**